### PR TITLE
[css-cascade-6] Add implicit scopes

### LINKED ISSUE #6606: [css-cascade-6] A way for <style> elements to scope to their parent element
https://drafts.csswg.org/css-cascade-6/#scoped-styles

The [explainer](https://css.oddbird.net/scope/explainer/#references--acknowledgements) references this issue: [Please bring back scoped styles](https://github.com/w3c/csswg-drafts/issues/3547).

Either there's something missing from the spec or I'm missing it, which is a way to have a `<style>` element that gets scoped to its own parent element.

```html
<div>
  <style scoped>p { color: red; }</style>
  <p>red</p>
</div>
```

`<style scoped>` would guarantee that the color only applied to children of the parent div.

I could do it by adding a unique ID or attribute:

```html
<div id="foo">
  <style>#foo p { color: red; }</style>
  <p>red</p>
</div>
```

… this is basically what Vue does at build time. But I want the same effect without a build step!

I think we'd want something like `@scope(:style-parent)`.

```html
<div>
  <style>@scope(:style-parent) { p { color: red; } }</style>
  <p>red</p>
</div>
```

Otherwise, even with `@scope` back in play, I'd still have to use a build tool (or something) to generate an ID for the upper bound of the scope.

### PR DESCRIPTION
**Pull request by andruud**
Resolves #6606 #7349

### CODE REVIEW

**Comment by andruud** [comment]
@mirisuzanne Please have a look.

I felt that I had to reorganize a bit here to make a natural place for "implicit" scopes to live.

Description of changes:

 - A scope should probably have a _single_ scoping root. However:
 - A single `@scope` rule can produce multiple (explicit) scopes. Or a single implicit scope.
 - Removed the circular or near-circular sentence "selectors are scoped to the given scope, with the :scope element being the scoping root" which Emilio complained about at some point (somewhere).
 - Avoid selector-4's definition of "scoped" for now, because it does not clarify what it means by "descendant", and does not account for scoping limits.
 - Beautify the grammar part of the spec.

Hopefully it's more or less to your liking.

File: css-cascade-6/Overview.bs [hunk 1]
```diff
@@ -351,17 +351,49 @@ Scoping Styles: the ''@scope'' rule</h4>
 
 	The syntax of the ''@scope'' rule is:
 
-	<pre class='prod'>
-	@scope (<<scope-start>>) [to (<<scope-end>>)]? {
+	<pre class="prod def">
+	@scope [ (<<scope-start>>) [to (<<scope-end>>)]? ]? {
```
Comment (mirisuzanne) [inline css-cascade-6/Overview.bs:355]: I think we should still allow scoping limits, even when the root is implicit?

```suggestion
	@scope [(<<scope-start>>)]? [to (<<scope-end>>)]? {
```

File: css-cascade-6/Overview.bs [hunk 2]
```diff
@@ -351,17 +351,49 @@ Scoping Styles: the ''@scope'' rule</h4>
...
+
+	* For each element matched by <<scope-start>>,
+		create a [=scope=] using that element as the [=scoping root=].
+	* For each [=scope=] created by the above:
+		* Collect all elements that are [=in scope=]
+			and that match <<scope-end>>,
+			using the [=scoping root=] as the '':scope'' element, then
+		* Set those elements as the [=scoping limits=].
+
+	[=Pseudo-elements=] cannot be [=scoping roots=] or [=scoping limits=];
```
Comment (mirisuzanne) [inline css-cascade-6/Overview.bs:378]: After moving this above in the previous PR, I moved it back down! 😅

**Review by mirisuzanne — COMMENTED** [review]
@andruud I think it's a big improvement, thanks. I suggested a change to the syntax production, allowing lower boundaries when the root is implicit. Otherwise, it looks good to me.

File: css-cascade-6/Overview.bs (hunk 2 above)
Comment (andruud) [reply css-cascade-6/Overview.bs:378]: :-)

**Comment by andruud** [comment]
> @andruud I think it's a big improvement, thanks. I suggested a change to the syntax production, allowing lower boundaries when the root is implicit.

Ah, right, that didn't occur to me. But then we need another fixup: PR: https://github.com/w3c/csswg-drafts/pull/8466

**Pull request merged**